---
title: "DNA-RNA"
output:
  flexdashboard::flex_dashboard:
    self_contained: true
    vertical_layout: fill
    theme: default
    logo: "/sibcb3/wangkailelab3/jinyu/coassay-dashboard/coassay_pipeline/dashboard/wang_lab_logo.png"
    css: "/sibcb3/wangkailelab3/jinyu/coassay-dashboard/custom.css"
    navbar:
      - { title: "WangLab", href: "https://wanglab.sibcb.ac.cn/", id: "about-link" }
params:
  outdir:
  id:
  ref:
  res:
  chip:
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
options(bitmapType="cairo")
library(dygraphs,lib.loc = "/sibcb1/wangkailelab1/pengjinyu/R/x86_64-pc-linux-gnu-library/4.4")
library(params,lib.loc = "/sibcb1/wangkailelab1/pengjinyu/R/x86_64-pc-linux-gnu-library/4.4")
library(knitr,lib.loc = "/sibcb1/wangkailelab1/pengjinyu/R/x86_64-pc-linux-gnu-library/4.4")
library(VennDiagram,lib.loc = "/sibcb1/wangkailelab1/pengjinyu/R/x86_64-pc-linux-gnu-library/4.4")

library(flexdashboard)
knitr::opts_chunk$set(
  cache = FALSE,
  cache.lazy = FALSE,
  tidy = TRUE,
  autodep=TRUE)
.libPaths(c("/sibcb/program/install/r-4.4/lib64/R/library"))
suppressPackageStartupMessages({
#library(magick)
library(dplyr)
library(Matrix)
library(gridExtra)
library(cowplot)
library(grid)
library(kableExtra)
library(ggpubr)
library(scater)
library(scran)
library(Seurat)
library(ggplot2)
library(scales)
library(copykat)
library(DoubletFinder)
library(rvest)
library(stringr)
library(matrixStats)
library(SingleR)
})
packages <- c("tidyverse", "fs", "shiny",
              "flexdashboard", "cowplot",
              "janitor", "ape", "ggsci", "plotly",
              "amap", "paletteer", "scales", "umap","uwot",
              "flsa", "BiocManager", "DT", "devtools", "Metrics","scquantum")
package.check <- lapply(packages, FUN = function(x) {
  if (!require(x, character.only = TRUE)) {
    install.packages(x, dependencies = TRUE,lib=.libPaths()[1],repos = "http://cran.us.r-project.org")
    suppressWarnings(library(x, character.only = TRUE))
  }
})

bioc_packages <- c("ggtree", "ComplexHeatmap")

lapply(bioc_packages, FUN = function(x) {
  if (!require(x, character.only = TRUE)) {
    BiocManager::install(x, update = FALSE)
    suppressWarnings(library(x, character.only = TRUE))
  }
})

git_packages<- c("Rphenograph")
lapply(git_packages, FUN = function(x) {
if(!require(x, character.only = TRUE)){
  devtools::install_github("JinmiaoChenLab/Rphenograph")
  library(x,character.only = TRUE)
}

})

theme_set(theme_cowplot())
#knitr::opts_knit$set(root.dir = paste0(params$outdir,"/final_result/"))
knitr::opts_knit$set(root.dir = paste0(params$outdir,"/"))

```

```{r genelists, include=FALSE}
####### Gene lists
# This section have the gene lists hard coded - they are used in later feature plots too.
# Please edit this section so the feature plots match the dot plots best.
# Changes to Gene List December 2022
#########################################################################
# Non Immune
# Endothelial
Endo_genes <- c("CDH5","VWF","PECAM1","ESAM","THBD","CD36", "CD34", "HEG1")
# Removed "LYZ"
# Added  "ENG","ESAM","THBD","CD36", "CD34", "HEG1""
# Epithelial
Epi_genes <- c("EPCAM","KRT5","KRT17","KRT6","KRT18","KRT19", "KRT8", "KRT14")
# Removed "EGFR"
# Added consider adding "KRT8", "KRT14", "KRT6B"
# Adipose
Adipo_genes <- c("APMAP","ADIPOQ","ADIPOQ-AS1","TPRA1")
# Removed
# Added
# Fibroblast Genes
Fibro_genes <- c("COL1A1","LUM","FBLN1", "DCN", "FBN1", "COL1A2", "PCOLCE")
# Removed "FAP",
# Added  "DCN", "FBN1", "COL1A2", "PCOLCE"
#non_immune_genes <- c(Endo_genes, Epi_genes, Adipo_genes,Fibro_genes)
#########################################################################
# Immune and lymphocytes
# T-cells
Tcell_genes <- c("CD3D","CD4","CD8A", "TRAC", "TRBC1", "TRBC2", "IL7R", "CD96")
# Removed
# Added   "TRAC", "TRBC1", "TRBC2", "IL7R", "CD96", "SYTL3"
# B-cells
Bcell_genes <- c("CD19","MS4A1","CD79A","CD79B","BANK1", "SEL1L3", "IGHM", "IGLC3")
# Removed
# Added   "BANK1", "SEL1L3", "IGHM", "IGLC3"
NKill_genes <-  c("NKG7","GNLY","CTSW")
# Removed "KLRF1"
# Added "KLRB1", "NCR1", "PRF1", "KLRD1"
Immun_genes <- c("PTPRC","SRGN","TYROBP") # double check
# Removed "LST1", "MNDA"
# Added "TYROBP"
#lymphocytes_cluster_genes <- c(Tcell_genes, Bcell_genes, NKill_genes,Immun_genes)
#########################################################################
# Myeloid
myeloid_genes <- c("FCER1G", "ITGAM", "FCGR3A") # the CD1c+ myeloid DCs, the CD141+ myeloid DCs and the CD303+ plasmacytoid DCs.
# Removed "Cd1c", "Cd141", "Cd303"
# Added "FCER1G", "ITGAM", "FCGR3A"
# Red Blood Cells
RedBld_genes <- c("HBA1","HBA2","HBB")
# Removed
# Added
# Macrophages
Macro_genes <- c("CD14","CD68","CD163","C1QA", "C1QB", "MRC1", "MSR1", "MS4A7")
# Removed "CD79B"
# Added  "C1QA", "C1QB", "MRC1", "MSR1", "MS4A7", "FCGR2A"
# Monocytes
Monoc_genes <- c("LST1","FCGR3A","AIF1", "MAFB", "MS4A6A", "ANPEP", "LYZ")
# Removed "CTSB"
# Added  "MAFB", "MS4A6A", "ANPEP", "LYZ"
Dendr_genes <- c("LILRA4", "IRF7", "CLEC9A", "CD1C", "TCF4")
# Removed "CTSV","SAMHD1","AXL","AREG",
# Added  "LILRA4", "IRF7", "CLEC9A", "CD1C", "TCF4"
#myeloid_cluster_genes <- c(myeloid_genes,RedBld_genes,Macro_genes, Monoc_genes,Dendr_genes)
#########################################################################
# something to add later would be user Input!
# user input genes
# user_input_genes <- read.delim()
# myPlotUserGeneList <-DotPlot(sample_com_filter, features = user_input_genes, assay='RNA',group.by = "seurat_clusters")  + coord_flip()+ggtitle("User Gene List")
# New list came from Aatish Thennavan confirm on "his" mouse data sets
```

```{r define_image_popup_function, results='asis', echo=FALSE}
cat('
<script>
function openBase64ImageInPopup(dataUrl, windowTitle, downloadFilename) {
  var popupHtml = `
  <!DOCTYPE html>
  <html>
  <head>
      <title>${windowTitle}</title>
      <style>
          body {
              margin: 0;
              padding: 10px; 
              height: 100vh;
              display: flex;
              justify-content: center;
              align-items: center;
              background-color: #f5f5f5; 
              box-sizing: border-box;
              overflow: hidden; 
          }
          img {
              max-width: 100%;
              max-height: 100%;
              width: auto;
              height: auto;
              object-fit: contain; 
              display: block;
              margin: auto;
          }
      </style>
  </head>
  <body>
      <img src="${dataUrl}" alt="${windowTitle}">
  </body>
  </html>`;


  var newWindow = window.open("", "_blank", "width=800,height=600,scrollbars=yes,resizable=yes");

  if (newWindow) {
    try {
      newWindow.document.open();
      newWindow.document.write(popupHtml);
      newWindow.document.close();

    } catch (e) {

      newWindow.close();
      alert("An error occurred while opening the image: " + e.message);
    }
  } else {

    alert("The pop-up window was blocked by the browser. Please allow pop-ups and try again.\\nOr, you can try saving the image manually.");

    var downloadLink = document.createElement("a");
    downloadLink.href = dataUrl;
    downloadLink.download = downloadFilename; 
    document.body.appendChild(downloadLink);


    if (document.createEvent) {
        var event = document.createEvent("MouseEvents");
        event.initEvent("click", true, true);
        downloadLink.dispatchEvent(event);
    } else {

        downloadLink.click();
    }

    document.body.removeChild(downloadLink);
  }
}
</script>
')
```

```{r load_dna}
# message
message("Rendering Dashboard")
# reading data
path=paste0(dirname(normalizePath(params$outdir)),"/")
message(path)
message(getwd())
dat_bin <- read.table(system(paste0("find ","/",params$outdir," -name \"*.bin.txt\" -type f"), intern = TRUE),sep="\t",header=T) %>% janitor::clean_names()
message(dim(dat_bin))
dat_seg <- read.table(system(paste0("find ","/",params$outdir," -name \"*.seg.txt\" -type f"), intern = TRUE),sep="\t",header=T) %>% janitor::clean_names()
dat_seg_cp <- dat_seg 
message(dim(dat_seg))

dat_stats <- read.table(system(paste0("find ","/",params$outdir," -name \"*.bam_metrics.txt\" -type f"), intern = TRUE),sep="\t",header=T) %>%
  janitor::clean_names() 
message(dim(dat_stats))

dat_all <- read.table(system(paste0("find ","/",params$outdir," -name \"*.raw_metrics.txt\" -type f"), intern = TRUE),sep="\t",header=T) %>%
  janitor::clean_names() %>% t() %>% as.data.frame()

# correcting names
dat_stats$sample <- make_clean_names(dat_stats$sample)


# dat_stats metrics
dat_stats_metrics <- dat_stats 
# reading ratio


file_path <- system(paste0("find ", params$outdir, " -name \"remove_cells.txt\" -type f"), intern = TRUE)

# 
if (file.exists(file_path) && file.info(file_path)$size > 0) {
  dat_rm <- read.table(file_path, sep = "\t", header = FALSE) %>%
    janitor::clean_names()
  dat_rm_n <- nrow(dat_rm)
} else {
  dat_rm_n <- 0
}


dat_rat <- read.table(system(paste0("find ","/",params$outdir," -name \"*.ratio.txt\" -type f"), intern = TRUE),sep="\t",header=T) %>%janitor::clean_names()
#dat_rat_cp <- dat_rat 

message(dim(dat_rat))
# dat_seg setup
dat_seg_s <- dat_seg 

dat_seg_t <- as.data.frame(t(dat_seg_s))
#dat_seg_t[1:5,1:5]

# n cells
n_cells <- dat_seg %>%ncol()

dat_ref <- read.table(system(paste0("find ","/",params$outdir," -name \"*.ref.txt\" -type f"), intern = TRUE),sep="\t",header=T)

# chipfile 

chipfile <- read.delim(params$chip)
filtered_cell <-  read.table(system(paste0("find ","/",params$outdir," -name \"*.bam_metrics.txt\" -type f"), intern = TRUE),sep="\t",header=T)
split_name <- stringr::str_split(filtered_cell$sample, "_")
new_df<- data.frame(sample =filtered_cell$sample,dispense = sapply(split_name, `[`, 5),
                    Row = as.numeric(gsub("r","",sapply(split_name, `[`, 3))) ,Col= 
                    as.numeric(gsub("c","",sapply(split_name, `[`, 4) ))  )



#false_pos 1
dispense_yes <- chipfile %>% filter(For.dispense=="yes") %>% left_join(.,new_df,by=c("Row","Col"))
wafergen <- read.csv("/sibcb3/wangkailelab3/jinyu/demultiplex/index/wafergen_list_DNA_RNA.txt",sep = " ")

false_pos <- new_df %>% filter(dispense == "no") 
if (nrow(false_pos) > 0) {
  false_pos$if_false_pos <- "yes"
  false_pos_table <- left_join(wafergen, false_pos, by = c("Row", "Col")) %>%
    mutate(
      if_false_pos = ifelse(is.na(if_false_pos), "no", if_false_pos)
    )
  barcode <- as.data.frame(table(false_pos_table$N7_Position,false_pos_table$S5_Position,false_pos_table$if_false_pos)) %>% filter(Var3=="yes")  %>%
  mutate(S5_Position=Var2,N7_Position=Var1,if_false_pos=Freq)

  barcode$S5_Position <- factor(barcode$S5_Position,levels = unique(wafergen$S5_Position)) 
  barcode$N7_Position <- factor(barcode$N7_Position,levels = unique(wafergen$N7_Position)) 
  barcode$if_false_pos <- factor(barcode$if_false_pos)

  p_false_pos <- ggplot(data = barcode, aes(x = N7_Position, y = S5_Position, fill = if_false_pos)) +  
    geom_tile(color = "white") +  
    scale_fill_manual(values = c("0" = "lightgrey", "1" = "red")) + 
    labs(x = "N7", y = "S5",fill="false pos") +
    theme_bw() +
    theme(
      legend.position = "bottom",
      axis.text.x = element_text(angle = 90, vjust = 0.5),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.text = element_text(size = 8),
      axis.title = element_text(size = 10),
      legend.title = element_text(size = 10),    
      legend.text = element_text(size = 10),
      legend.key.size = unit(0.3, "cm"),
      legend.key.width = unit(0.3, "cm"),    
      legend.key.height = unit(0.3, "cm"),
      aspect.ratio = 1
    )
} else {
  false_pos_table <- wafergen %>%
    mutate(if_false_pos = "no")
  barcode <- as.data.frame(table(false_pos_table$N7_Position,false_pos_table$S5_Position,false_pos_table$if_false_pos)) %>%
  mutate(S5_Position=Var2,N7_Position=Var1,if_false_pos=0)

  barcode$S5_Position <- factor(barcode$S5_Position,levels = unique(wafergen$S5_Position)) 
  barcode$N7_Position <- factor(barcode$N7_Position,levels = unique(wafergen$N7_Position)) 
  barcode$if_false_pos <- factor(barcode$if_false_pos)

  p_false_pos <- ggplot(data = barcode, aes(x = N7_Position, y = S5_Position, fill = if_false_pos)) +  
    geom_tile(color = "white") +  
    scale_fill_manual(values = c("0" = "lightgrey", "1" = "red")) + 
    labs(x = "N7", y = "S5",fill="false pos") +
    theme_bw() +
    theme(
      legend.position = "bottom",
      axis.text.x = element_text(angle = 90, vjust = 0.5),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.text = element_text(size = 8),
      axis.title = element_text(size = 10),
      legend.title = element_text(size = 10),    
      legend.text = element_text(size = 10),
      legend.key.size = unit(0.3, "cm"),
      legend.key.width = unit(0.3, "cm"),    
      legend.key.height = unit(0.3, "cm"),
      aspect.ratio = 1
    )
}

#false neg
false_neg <- dispense_yes %>% filter(For.dispense == "yes" & is.na(sample)) 
false_neg$if_false_neg <- "yes"
false_neg_table <- left_join(wafergen,false_neg,by=c("Row","Col")) %>%  mutate (
  if_false_neg = ifelse(is.na(if_false_neg), "no",if_false_neg)
)
barcode <- as.data.frame(table(false_neg_table$N7_Position,false_neg_table$S5_Position,false_neg_table$if_false_neg)) %>% filter(Var3=="yes") %>%
  mutate(S5_Position=Var2,N7_Position=Var1,if_false_neg=Freq)

barcode$S5_Position <- factor(barcode$S5_Position,levels = unique(wafergen$S5_Position)) 
barcode$N7_Position <- factor(barcode$N7_Position,levels = unique(wafergen$N7_Position)) 
barcode$if_false_neg <- factor(barcode$if_false_neg)

p_false_neg <- ggplot(data = barcode, aes(x = N7_Position, y = S5_Position, fill = if_false_neg)) +  
  geom_tile(color = "white") +  
  scale_fill_manual(values = c("0" = "lightgrey", "1" = "red")) + 
  labs(x = "N7", y = "S5",fill="false neg") +
  theme_bw() +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 90, vjust = 0.5),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text = element_text(size = 8),
    axis.title = element_text(size = 10),
    legend.title = element_text(size = 10),    
    legend.text = element_text(size = 10),
    legend.key.size = unit(0.3, "cm"),
    legend.key.width = unit(0.3, "cm"),    
    legend.key.height = unit(0.3, "cm"),
    aspect.ratio = 1
  )



raw_data <- read.table(system(paste0("find ","/",params$outdir," -name \"*.raw_metrics.txt\" -type f"), intern = TRUE),sep="\t",header=T) %>% 
  t() %>% as.data.frame()
raw_data$total <- rowSums(raw_data)
split_name <- stringr::str_split(rownames(raw_data), "_")
raw_data_df<- data.frame(sample =rownames(raw_data),read_num = raw_data$total,
                    Row = as.numeric(gsub("r","",sapply(split_name, `[`, 3))) ,Col= 
                      as.numeric(gsub("c","",sapply(split_name, `[`, 4) ))  )

reads_num_table <- left_join(wafergen,raw_data_df,by=c("Row","Col")) 
reads_num_table_cell <- left_join(reads_num_table,new_df,by=c("Row","Col")) %>% mutate(
  real_cell_reads = read_num,
  non_cell_reads = read_num
) %>% mutate(
   real_cell_reads = ifelse(is.na(dispense), NA,real_cell_reads),
   non_cell_reads = ifelse(is.na(dispense), non_cell_reads,NA)
)


p_real_reads_num <- ggplot(data = reads_num_table_cell, aes(x = N7_Position, y = S5_Position, fill = real_cell_reads)) +  
  geom_tile(color = "white") +  
  scale_fill_gradient(low = "#ffe5ec", high = "red",na.value = "white") + 
  labs(x = "N7", y = "S5",fill="reads num/well") +
  theme_bw() +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 90, vjust = 0.5),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text = element_text(size = 8),
    axis.title = element_text(size = 10),
    legend.title = element_text(size = 10),    
    legend.text = element_text(size = 10),
    legend.key.size = unit(0.3, "cm"),
    legend.key.width = unit(2, "cm"),    
    legend.key.height = unit(0.3, "cm"),
    aspect.ratio = 1
  )
p_non_reads_num <- ggplot(data = reads_num_table_cell, aes(x = N7_Position, y = S5_Position, fill = non_cell_reads)) +  
  geom_tile(color = "white") +  
  scale_fill_gradient(low = "#e8eef1", high = "#1e3d58",na.value = "white") + 
  labs(x = "N7", y = "S5",fill="reads num/well") +
  theme_bw() +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 90, vjust = 0.5),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text = element_text(size = 8),
    axis.title = element_text(size = 10),
    legend.title = element_text(size = 10),    
    legend.text = element_text(size = 10),
    legend.key.size = unit(0.3, "cm"),
    legend.key.width = unit(2, "cm"),    
    legend.key.height = unit(0.3, "cm"),
    aspect.ratio = 1
  )
dir.create("final_result/figures", recursive = T)
ggsave(filename =paste0(params$outdir, "/final_result/figures/real_reads_num_heatmap.png"), plot = p_real_reads_num, width = 8, height = 8)
ggsave(filename = paste0(params$outdir,"/final_result/figures/non_cell_reads_on_chip_heatmap.png"), plot = p_non_reads_num, width = 8, height = 8)
ggsave(filename = paste0(params$outdir,"/final_result/figures/false_pos_on_chip_heatmap.png"), plot = p_false_pos, width = 8, height = 8)
ggsave(filename = paste0(params$outdir,"/final_result/figures/false_neg_on_chip_heatmap.png"), plot = p_false_neg, width = 8, height = 8)
```

```{r clustering, include = FALSE}
if (n_cells < 20) {
  umap_n_neighbors <- n_cells - 1
} else umap_n_neighbors <- 20

set.seed(42)
dat_umap <- uwot::umap(dat_seg_t, 
                       metric = "manhattan",
                       n_threads = 20,
                       min_dist = 0.01,
                       n_neighbors = umap_n_neighbors)
  
umap_df <- as.data.frame(dat_umap) %>% 
  dplyr::rename("umap1" = "V1",
                "umap2" = "V2")
rownames(umap_df) <- rownames(dat_seg_t)

if (n_cells < 20) {
  k_param <- n_cells - 1
} else k_param <- 20
  
pheno <- Rphenograph::Rphenograph(umap_df, k = k_param)
cl <- tibble(cell = rownames(umap_df),
             cluster = igraph::membership(pheno[[2]])) %>% 
  arrange(cluster)

umap_df <- umap_df[cl$cell,]

# getting number of needed colors for cl
n_colors <- cl$cluster %>% unique() %>% length()
# ordering cells
dat_seg_order <- dat_seg_t[cl$cell,]

# setting colors
row_col <- setNames(colorRampPalette(paletteer_d("pals::alphabet", 26))(n_colors),
                    levels(as.factor(cl$cluster)))
```

# [DNA Summary]{.my-dna}

## Column {data-width="100"}

### Cells Sequenced {.colored}

```{r}
total <- dat_rm_n + n_cells
valueBox(total,icon = "circle",
           color = "#729EA1")
```

### Cells removed

```{r}

valueBox(dat_rm_n, icon = "ion-trash-a",
         color = "#EC9192")
```

### Dispense Yes

```{r}
ptg_dna <- round(n_cells/length(dispense_yes$dispense),2) *100
paste0(length(dispense_yes$dispense),"(",ptg_dna,"%)") %>%
  valueBox(icon = "circle",
           color = "#DFBE99")
```

### Cells passed QC {.colored}

```{r}
valueBox(n_cells,
        icon = "ion-android-done-all",
        color = "#B5BD89")
```

### Total Reads (All PF Reads) {.colored}

```{r tot_reads_val_box_all}
sum(dat_stats$reads_total) %>% 
  valueBox(icon = "circle",
           color = "#ced3d7")
```

### Total Reads Per Cell(Median) {.colored}

```{r tot_reads_val_box}
median(dat_stats$reads_total) %>% 
  valueBox(icon = "circle",
           color = "#ced3d7")
```

### Kept Reads Per Cell (Median) {.colored}

```{r kep_reads_val_box}
median(dat_stats$reads_assigned_bins) %>% 
  valueBox(icon = "circle",
           color = "#ced3d7")
```

### Bin Count (Median) {.colored}

```{r bincount_val_box}
medians <- apply(dat_bin, 2, median, na.rm = TRUE)
median(medians) %>% 
  valueBox(icon = "circle",
           color = "#ced3d7")
```

### PCR Duplicates (Median) {.colored}

```{r calculating_pcr_duplicates}
dups <- median(round(dat_stats$percentage_duplicates*100,2))
valueBox(paste0(dups,"%"),icon = "circle",
           color = "#ced3d7")

```

### Median Ploidy (scquantum)

```{r infering_ploidies}

all.values <- apply(dat_bin, 2, ploidy.inference) #, chrom=dat_bin$chrom)
df<-data.frame(sample_name=names(all.values),
           ploidy=sapply(all.values, function(x) x$ploidy),
           peak.height=sapply(all.values, function(x) x$peak_height)) %>%
            dplyr::mutate(Cell = paste(sample_name,
                              ": ",
                             ploidy,
                              sep = ""))
df$ploidy[is.na(df$ploidy)]<-0
ploidy_median <- median(df$ploidy)
round(ploidy_median,2) %>% valueBox(icon = "circle",
           color = "#ced3d7")

```

### Clusters

```{r n_clusters}
cl$cluster %>% unique() %>% length() %>% 
  valueBox(icon = "circle",
           color = "#ced3d7")
```

### Reference

```{r}
 params$ref %>%
  valueBox(icon = "circle",
           color = "#ced3d7")
```

### Resolution

```{r}
 params$res %>%
  valueBox(icon = "circle",
           color = "#ced3d7")
```

### False Positive

```{r}
 sum(new_df$dispense=="no") %>%
  valueBox(icon = "circle",
           color = "#ced3d7")
```

### False Negative

```{r}
 sum(is.na(dispense_yes$dispense)==TRUE) %>%
  valueBox(icon = "circle",
           color = "#ced3d7")
```

## Column {data-width="1000"}

### Heatmap

```{r making_heatmap,fig.width= 10,fig.height=14}
# chromosome annotation top bar
# getting the vector of chrom lengths

dat_ref$chr <- factor(dat_ref$chr,levels = unique(dat_ref$chr))
chr_lengths <- dat_ref %>%
  dplyr::select(abspos, chr, start) %>% 
  group_by(chr) %>%
  dplyr::summarise(n = n()) %>%
  dplyr::pull(n)


chr_binary <- rep(c(2,1), 12)
chr <- data.frame(chr = rep.int(x = chr_binary, times = chr_lengths))
# getting lengths for chr numbers annotation
chr_rl_c <- c(1, cumsum(chr_lengths))
# creating a data frame to calculate rowMeans
chr_df <-  data.frame(a = chr_rl_c[1:length(chr_rl_c)-1],b= chr_rl_c[2:length(chr_rl_c)])
chr_l_means <- round(rowMeans(chr_df))
chrom.names <- c(1:22,"X", "Y")
# creating the vector for chr number annotations
v <- vector(mode = "character",length = cumsum(chr_lengths)[length(chr_lengths)])
v[chr_l_means] <- chrom.names
v[is.na(v)] <- ""
chr_bar <- HeatmapAnnotation(chr_text = anno_text(v, 
                                                  gp = gpar(fontsize = 8)),
                             df = chr,
                             show_legend = FALSE,
                             which = "column",
                             col = list(chr = c("1" = "grey88", "2" = "black"))
)

#Peak=rep("Peak_D",nrow(dat_seg_order))
#Peak[grep(rownames(dat_seg_order),pattern = "_A_",ignore.case = T)]="Peak_A"
#Peak[grep(rownames(dat_seg_order),pattern = "A_",ignore.case = T)]="Peak_A"
#Peak[grep(rownames(dat_seg_order),pattern = "Ctrl",ignore.case = T)]="Ctrl"

Cells=rep("preTX",nrow(dat_seg_t))
Cells[grep(rownames(dat_seg_t),pattern = "preTX",ignore.case = T)]="MIDTX"
col<-c("blue","orange")
names(col)<-c("MIDTX", "preTX")

ht <- Heatmap(as.matrix(log2(dat_seg_t+1e-3)),
              column_title = params$id,
              column_title_gp = gpar(fontsize = 30, fontface = "bold"),
              cluster_columns = FALSE,
              cluster_rows = TRUE,
              show_row_names = FALSE,
              show_column_names = FALSE,
              use_raster = TRUE,
              raster_quality = 5,
              top_annotation = chr_bar,
	            col = circlize::colorRamp2(breaks = c(-2,0,2),c("dodgerblue3", "white", "firebrick3")),
              heatmap_legend_param = list(title = "Log2 (Ratio)", title_gp = gpar(fontsize = 12, fontface = "bold"),labels_gp = gpar(fontsize = 12)))


draw(ht,heatmap_legend_side = "right")
```

## Column {data-width="500"}

### False Positive Well Heatmap {data-height="500"}

```{r falsepos_heatmap_dna, results='asis'}
img_file_path <- paste0(params$outdir,"/final_result/figures/false_pos_on_chip_heatmap.png") 
if (!file.exists(img_file_path)) {
  cat('<p style="color: red; font-weight: bold;">ERROR: Image file not found at ', img_file_path, '</p>')
} else {
  img_data <- readBin(img_file_path, what = "raw", n = file.info(img_file_path)$size)
  img_base64 <- base64enc::base64encode(img_data)
  img_data_uri <- paste0("data:image/png;base64,", img_base64)
  cat('<div style="text-align: center; margin: 5px 0;">')
  cat(sprintf(
    '<img src="%s" alt="False Pos Well Heatmap" id="false_pos_on_chip_heatmap_rna" style="max-width: 100%%; height: auto; cursor: pointer; display: block; margin: 0 auto;" onclick="openBase64ImageInPopup(\'%s\', \'False Pos Well Heatmap\', \'false_pos_on_chip_heatmap.png\')" />',
    img_data_uri, img_data_uri
  ))
  cat('</div>')
}
#knitr::include_graphics("final_result/figures/false_pos_on_chip_heatmap.png")
```

### Filtered-cells Reads Num Heatmap {data-height="500"}

```{r reads_num_heatmap_dna, results='asis'}
img_file_path <- paste0(params$outdir,"/final_result/figures/real_reads_num_heatmap.png") 
if (!file.exists(img_file_path)) {
  cat('<p style="color: red; font-weight: bold;">ERROR: Image file not found at ', img_file_path, '</p>')
} else {
  img_data <- readBin(img_file_path, what = "raw", n = file.info(img_file_path)$size)
  img_base64 <- base64enc::base64encode(img_data)
  img_data_uri <- paste0("data:image/png;base64,", img_base64)
  cat('<div style="text-align: center; margin: 5px 0;">')
  cat(sprintf(
    '<img src="%s" alt="Filtered-cells Reads Num Heatmap" id="real_reads_num_heatmap_rna" style="max-width: 100%%; height: auto; cursor: pointer; display: block; margin: 0 auto;" onclick="openBase64ImageInPopup(\'%s\', \'Filtered-cells Reads Num Heatmap\', \'real_reads_num_heatmap.png\')" />',
    img_data_uri, img_data_uri
  ))
  cat('</div>')
}
#knitr::include_graphics("final_result/figures/real_reads_num_heatmap.png")
```

## Column {data-width="500"}

### False Negative Well Heatmap {data-height="500"}

```{r falseneg_heatmap_dna, results='asis'}
img_file_path <- paste0(params$outdir,"/final_result/figures/false_neg_on_chip_heatmap.png") 
if (!file.exists(img_file_path)) {
  cat('<p style="color: red; font-weight: bold;">ERROR: Image file not found at ', img_file_path, '</p>')
} else {
  img_data <- readBin(img_file_path, what = "raw", n = file.info(img_file_path)$size)
  img_base64 <- base64enc::base64encode(img_data)
  img_data_uri <- paste0("data:image/png;base64,", img_base64)
  cat('<div style="text-align: center; margin: 5px 0;">')
  cat(sprintf(
    '<img src="%s" alt="False Negative Well Heatmap" id="falseneg_heatmap_dna" style="max-width: 100%%; height: auto; cursor: pointer; display: block; margin: 0 auto;" onclick="openBase64ImageInPopup(\'%s\', \'False Negative Well Heatmap\', \'falseneg_heatmap_dna.png\')" />',
    img_data_uri, img_data_uri
  ))
  cat('</div>')
}

#knitr::include_graphics("final_result/figures/false_neg_on_chip_heatmap.png")
```

### Non-cells Reads Num Heatmap {data-height="500"}

```{r non_cell_reads_num_heatmap_dna, results='asis'}
img_file_path <- paste0(params$outdir,"/final_result/figures/non_cell_reads_on_chip_heatmap.png")
if (!file.exists(img_file_path)) {
  cat('<p style="color: red; font-weight: bold;">ERROR: Image file not found at ', img_file_path, '</p>')
} else {
  img_data <- readBin(img_file_path, what = "raw", n = file.info(img_file_path)$size)
  img_base64 <- base64enc::base64encode(img_data)
  img_data_uri <- paste0("data:image/png;base64,", img_base64)
  cat('<div style="text-align: center; margin: 5px 0;">')
  cat(sprintf(
    '<img src="%s" alt="Non-cells Reads Num Heatmap" id="non_cell_reads_num_heatmap_dna" style="max-width: 100%%; height: auto; cursor: pointer; display: block; margin: 0 auto;" onclick="openBase64ImageInPopup(\'%s\', \'Non-cells Reads Num Heatmap\', \'non_cell_reads_num_heatmap_dna.png\')" />',
    img_data_uri, img_data_uri
  ))
  cat('</div>')
}
#knitr::include_graphics("final_result/figures/non_cell_reads_on_chip_heatmap.png")
```

# [DNA QC]{.my-dna}

## Column {data-width="1000"}

### Quality, Bin Count,Reads and Duplication

```{r Reads_and_Duplication,fig.height=8,fig.width=12}
message("Calculating RMSE")
barfill <- "#4271AE"
rmses <- vector()

for (i in 1:ncol(dat_seg)) {
  actual <- as_tibble(dat_seg[, i]) %>% dplyr::pull()
  predicted <- as_tibble(dat_rat[, i]) %>% dplyr::pull()

  rmses[i] <- rmse(actual, predicted)
  names(rmses)[i] <- names(dat_seg)[i]
}

rmses_df <- enframe(rmses,
  name = "sample_name",
  value = "rmse"
)

# cv
message("Calculating CV")
cv <- vector()

for(i in 1:ncol(dat_seg)) {
  seg_n_probes <- rle(pull(as_tibble(dat_seg[,i])))$lengths
  ratio_split <- split(pull(as_tibble(dat_rat[,i])), 
                       as.factor(rep.int(1:length(seg_n_probes), seg_n_probes)))
  cv[i] <- lapply(ratio_split, function(x) sd(x)/mean(x)) %>% unlist() %>% sum()
  names(cv)[i] <- names(dat_seg)[i]
}

cv_df <- enframe(cv,
                      name = "sample_name",
                      value = "cv"
)

cv_rmse_df <- dplyr::inner_join(rmses_df, cv_df)

p <- cv_rmse_df %>% 
  ggplot(aes(cv, rmse)) +
  geom_point(aes(label = sample_name), size = 0.05, alpha = 0.8) +
  geom_smooth(method = "lm") +
  #viridis::scale_color_viridis(discrete = TRUE) +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10)) +
  scale_y_continuous(breaks = scales::pretty_breaks(n = 10)) +
  xlab("Coefficient of Variation") + 
  ylab("RMSE")+theme(axis.title.x = element_text(size=12),
                     axis.title.y = element_text(size=12,),
                     axis.text.x = element_text(size=10),
                     axis.text.y = element_text(size=10))

#p <- ggplotly(p)

p_rmse <- dplyr::inner_join(dat_stats, cv_rmse_df,by=c("sample"="sample_name")) %>% 
  ggplot(aes(reads_assigned_bins, rmse)) + 
  geom_point(size = 0.05, alpha = 0.8) +
  geom_smooth() +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10)) + 
  scale_y_continuous(breaks = scales::pretty_breaks(n = 10)) + 
 # scale_color_viridis_c() +
  theme(axis.text.x = element_text(angle = 90)) +
  ylab("RMSE")+theme(axis.title.x = element_text(size=12),
                     axis.title.y = element_text(size=12,),
                     axis.text.x = element_text(size=10),
                     axis.text.y = element_text(size=10))
  

p_bin_count <- dat_bin %>% 
  purrr::map_dfr(median) %>% 
  gather(key = "cell",
         value = "bin_median") %>% 
  dplyr::filter(bin_median > 1) %>% 
  ggplot(aes(x = "", y = bin_median)) +
  geom_violin(fill = "#fbd680", alpha = 0.8) + 
  geom_jitter(size=0.05, alpha = 0.6)+
  geom_boxplot(width = 0.2, outlier.size = 0.5) + 
  ggtitle("GC normalized median bin count")+
  xlab("") + 
  ylab("") + 
  theme(
    axis.text.x = element_blank(),
    plot.title = element_text(size = 11),
    axis.text.y = element_text(size=8)
  )


processed_data <- dat_stats %>% 
  gather(key = "parameter", value = "value", -sample) %>% 
  dplyr::filter(parameter %in% c("reads_total", "reads_assigned_bins"), value > 500)

p_reads_total <- processed_data %>%
  dplyr::filter(parameter == "reads_total") %>% 
  ggplot(aes(x = parameter, y = value)) +
  geom_violin(fill = "#fbd680", color = "black", alpha = 0.8) + 
  geom_jitter(size=0.05, alpha = 0.6)+
  geom_boxplot(width = 0.2, outlier.size = 0.5) +
  scale_y_continuous(labels = scientific_format())+
  ggtitle("Reads_total Count")+
  theme(
    axis.title.x = element_blank(), 
    axis.title.y = element_blank(),  
    axis.text.x = element_blank(),
    plot.title = element_text(size = 11),
    axis.text.y = element_text(size=10)
  ) 
  
p_reads_assigned <- processed_data %>%
  dplyr::filter(parameter == "reads_assigned_bins") %>% 
  ggplot(aes(x = parameter, y = value)) +
  geom_violin(fill = "#fbd680", color = "black", alpha = 0.8) + 
  geom_jitter(size=0.05, alpha = 0.6)+
  geom_boxplot(width = 0.2, outlier.size = 0.5) +
  scale_y_continuous(labels = scientific_format())+
  ggtitle("Reads_assigned_bins Count")+
  theme(
    axis.title.x = element_blank(), 
    axis.title.y = element_blank(),  
    axis.text.x = element_blank(),
    plot.title = element_text(size = 11),
    axis.text.y = element_text(size=10)
  ) 
  

p1_violin <- dat_stats %>% 
  ggplot(aes(x = "", y = percentage_duplicates)) +
  geom_violin(fill = "#fbd680", color = "black", alpha = 0.8) + 
  geom_jitter(size=0.05, alpha = 0.6)+
  geom_boxplot(width = 0.2, outlier.size = 0.5) +
  ggtitle("Duplication Rate")+
  theme(
    axis.title.x = element_blank(), 
    axis.title.y = element_blank(),  
    axis.text.x = element_blank(),
    plot.title = element_text(size = 11),
    axis.text.y = element_text(size=10)
  ) 
  
p2 <- dat_stats %>% mutate(duplicates=round(reads_duplicates,2),total_reads=round(reads_total/1e+6,2)) %>% 
  ggplot(aes(total_reads, duplicates)) +
  geom_point(aes(label = sample), size = 0.05, alpha = 0.8) +
  #geom_smooth(method = "lm") +
  #viridis::scale_color_viridis(discrete = TRUE) +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10)) +
  scale_y_continuous(breaks = scales::pretty_breaks(n = 10)) +
  xlab("Read count per cell(M)") + 
  ylab("duplication count")+xlim(0,2)+theme(axis.title.x = element_text(size=12),
                     axis.title.y = element_text(size=12),
                     axis.text.x = element_text(size=10),
                     axis.text.y = element_text(size=10))

p_line1 <- plot_grid(p,p_rmse,p2,nrow=1,align = "h",axis = "x")
p_line2 <- plot_grid(p_reads_total,p_reads_assigned,p1_violin,p_bin_count,nrow=1,align = "h",axis = "x")
plot_grid(p_line1,p_line2,nrow = 2)  
```

## Column {data-width="600"}

### UMAP {data-hight="500"}

```{r generating_umap}
p_umap <- umap_df %>% 
  rownames_to_column(var = "Cell") %>%
  dplyr::mutate(cluster = cl$cluster) %>% 
  ggplot() +
  geom_point(aes(x = umap1, 
                 y = umap2, 
                 color = as.factor(cluster),
                 label = Cell),
             show.legend = FALSE) +
  #scale_color_manual(values = row_col) +
  theme(legend.position = "none",
        panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank()) 
p_umap
#ggplotly(p_umap, tooltip = "Cell")
```

### Ploidy Inference {data-hight="500"}

```{r draw_ploidies}



# obtaining clusters labels to use as colors, matching with cell name and plotting
umap_cl_result <- umap_df %>% rownames_to_column("sample_name") %>%  mutate(cluster = cl$cluster) %>% dplyr::select(sample_name, cluster) 

dat_ploidy_cl <- dplyr::inner_join(df,umap_cl_result, by = "sample_name")

p_ploidy <- ggplot(dat_ploidy_cl,aes(x = ploidy, 
                 y = peak.height, label = Cell)) + 
  geom_point(color = as.factor(dat_ploidy_cl$cluster)) +
 # scale_color_manual(values = row_col) + 
  scale_x_continuous(breaks = scales::pretty_breaks()) +
  scale_y_continuous(breaks = scales::pretty_breaks()) +
  xlab("Inferred ploidy") +
  ylab("Peak Height (Confidence)") + 
  labs(color = "Cluster")

#ggplotly(p_ploidy, tooltip = "Cell")
p_ploidy
```

# [Reads information]{.my-dna}

## Column {.tabset .tabset-fade .tabset-pills}

### Kept cells information

```{r}
dat_bin_t <-apply(dat_bin, 2, mean, na.rm = TRUE) %>% as.data.frame() %>% rownames_to_column("sample")
names(dat_bin_t) <- c("sample","mean_normalized_bincount")
dat_bin_t$mean_normalized_bincount <- round(dat_bin_t$mean_normalized_bincount,2)
dt_dat_stats <- left_join(dat_stats,dat_bin_t,by="sample") 
dt_dat_stats <- dt_dat_stats %>% column_to_rownames("sample") 

DT::datatable(dt_dat_stats, options = list(pageLength = 1000))
```

### Removed cells information

```{r}
dat_all$reads_total <- rowSums(dat_all)
dt_dat_stats_rm <- dat_all %>% rownames_to_column("sample") %>% filter(!sample %in% dat_stats$sample) %>% column_to_rownames("sample") %>%
  mutate(percentage_duplicates=round(reads_duplicates/reads_total,2))

DT::datatable(dt_dat_stats_rm, options = list(pageLength = 1000))
```

# [Example Profiles of Ratio Plots]{.my-dna}

```{r making_ratio_plots}
plots<-dir_ls("final_result/ratio_plots_CN")
```

## Column {.tabset .tabset-fade}

### 1-10

```{r,out.width = "100%"}
knitr::include_graphics(plots[1:10])
```

### 11-20

```{r ,out.width = "100%"}
knitr::include_graphics(plots[11:20])
```

### 21-30

```{r,out.width = "100%"}
knitr::include_graphics(plots[21:30])
```

# [Phylogenetics]{.my-dna}

## Column

### Neighbor-Joining tree

```{r running_phylogenetics}
tree <- ape::nj(
  amap::Dist(dat_seg_order,
             method = "manhattan",
             nbproc = 40)
)

# obtaining colors
list_samples <- split(rownames(dat_seg_order),
                      sort(cl$cluster))
tree <- ggtree::groupOTU(tree, list_samples)

ggtree(ladderize(tree),
  ladderize = FALSE,
  size = .1) +
  geom_tippoint(aes(color = group), size = .7) +
  scale_colour_manual(values = c("black", row_col)) +
  theme(legend.position = "right") +
  labs(color = "Cluster")


```

# [RNA QC]{.my-rna}

## Column {data-width="100"}

```{r rna_loading,message=FALSE,warning = FALSE}

line_sum <- read.delim(paste0(params$outdir,"/rawdata/line_sum.txt"),sep = " ",header = F)
colnames(line_sum) <- c("cell", "reads")
line_sum_new <- line_sum %>%
  mutate(cell_names = basename(cell)) %>% select(cell_names,reads)
line_sum_new$cell_names <- sub("\\.fq\\.gz$", "", line_sum_new$cell_names)

counts_matrix <- read.delim(paste0(params$outdir,"/data/counts_matrix.txt"),row.names = 1)
counts_matrix <- counts_matrix %>% select(-X) 
seurat_obj <- CreateSeuratObject(counts = counts_matrix,project = params$id)
seurat_obj_mtx <- as.data.frame(seurat_obj@meta.data) 
seurat_obj_mtx$cellnames <- rownames(seurat_obj_mtx)
wafergen <- read.csv("/sibcb3/wangkailelab3/jinyu/demultiplex/index/wafergen_list_DNA_RNA.txt",sep = " ")
wafergen$rna_name <- paste0(wafergen$rna_n7_cr,wafergen$rna_s5)
seurat_obj_mtx_chip <- left_join(wafergen,seurat_obj_mtx,by = c("rna_name"="cellnames"))
chip_file <- read.csv(params$chip,sep = "\t")
seurat_obj_mtx_chip <- left_join(seurat_obj_mtx_chip,chip_file,by=c("Row","Col")) 
seurat_obj_mtx_chip <- seurat_obj_mtx_chip %>% filter(Sample == "sample")
false_pos <- seurat_obj_mtx_chip %>% filter(For.dispense == "no" & !is.na(orig.ident)) 
false_neg <- seurat_obj_mtx_chip %>% filter(For.dispense == "yes" & is.na(orig.ident)) 

passed_cell <- left_join(seurat_obj_mtx,line_sum_new,by=c("cellnames"="cell_names"))


if(length(false_pos$sample) >0){
  false_pos$if_false_pos <- "yes"
  false_pos_table <- left_join(wafergen,false_pos,by = c("rna_name")) %>%  mutate (
  if_false_pos = ifelse(is.na(if_false_pos), "no",if_false_pos)
)
}else{
  false_pos_table <-false_pos
}

filt_cells_names_arcDR <- seurat_obj_mtx %>% dplyr::filter(!(cellnames %in% false_pos$rna_name))
mtx_p1 <- subset(seurat_obj, cells = filt_cells_names_arcDR$cellnames)
mtx_p1_meta <- as.data.frame(mtx_p1@meta.data)
mtx2 <- mtx_p1
mtx2[["percent.mt"]] <- PercentageFeatureSet(mtx2, pattern = "^MT-")


sample.gene <- apply(counts_matrix, 2, function(x) sum(x>0))


sample_df<-data.frame(nUMI=colSums(as.matrix(counts_matrix)),nGene=sample.gene)
sample_UMI_Gene<-ggplot(sample_df, aes(nUMI, nGene))+ geom_point(size=0.5)+ scale_x_continuous(breaks=seq(0,200000,100000),limits=c(0,200000))


```

### Cells Sequenced {.colored}

```{r}
total <- length(line_sum$cell)
valueBox(total,icon = "circle",
           color = "#729EA1")
```

### Cells removed

```{r}
valueBox(total-length(seurat_obj_mtx$orig.ident), 
         icon = "ion-trash-a",
         color = "#EC9192")
```

### Dispense Yes

```{r}
ptg <- round(length(seurat_obj_mtx$orig.ident)/length(dispense_yes$dispense),2) *100
paste0(length(dispense_yes$dispense),"(",ptg,"%)") %>%
  valueBox(icon = "circle",
           color = "#DFBE99")
```

### Cells passed QC {.colored}

```{r}
valueBox(length(seurat_obj_mtx$orig.ident),
        icon = "ion-android-done-all",
        color = "#B5BD89")
```

### Total Reads (All PF Reads) {.colored}

```{r }
sum(passed_cell$reads/4) %>% 
  valueBox(icon = "circle",
           color = "#ced3d7")
```

### Total Reads Per Cell(Median) {.colored}

```{r}
median(passed_cell$reads/4) %>% 
  valueBox(icon = "circle",
           color = "#ced3d7")
```

### False Positive

```{r}
 length(false_pos$Sample) %>%
  valueBox(icon = "circle",
           color = "#ced3d7")
```

### False Negative

```{r}
 length(false_neg$Sample) %>%
  valueBox(icon = "circle",
           color = "#ced3d7")
```

## Column

### Gaussian Mixture Model {data-height="500"}

```{r Gaussian_Mixture_Model,results='asis'}
img_file_path <- paste0(params$outdir,"/rawdata/Gaussian_Mixture_Model.png") 
if (!file.exists(img_file_path)) {
  cat('<p style="color: red; font-weight: bold;">ERROR: Image file not found at ', img_file_path, '</p>')
} else {

  img_data <- readBin(img_file_path, what = "raw", n = file.info(img_file_path)$size)
  img_base64 <- base64enc::base64encode(img_data)
  img_data_uri <- paste0("data:image/png;base64,", img_base64)
  cat('<div style="text-align: center; margin: 5px 0;">')
  cat(sprintf(
    '<img src="%s" alt="Gaussian Mixture Model" id="Gaussian_Mixture_Model" style="max-width: 100%%; height: auto; cursor: pointer; display: block; margin: 0 auto;" onclick="openBase64ImageInPopup(\'%s\', \'Gaussian Mixture Model\', \'Gaussian_Mixture_Model.png\')" />',img_data_uri, img_data_uri
  ))
  cat('</div>')
}
#knitr::include_graphics("rawdata/rna_process/FILE/Gaussian_Mixture_Model.png")
```

### RNA QC

```{r sample_4, message=FALSE,warning = FALSE,include=FALSE}
p1_violin <- ggplot(mtx2@meta.data, aes(x=orig.ident, y=nFeature_RNA))+geom_violin(fill="coral")+ylab("nGene")+xlab("")+
  geom_jitter(size=0.5)+theme(axis.text.x = element_blank(), legend.position = "none")
p2_violin <- ggplot(mtx2@meta.data, aes(x=orig.ident, y=nCount_RNA))+geom_violin(fill="coral")+ylab("nUMI")+xlab("")+
  geom_jitter(size=0.5)+theme(axis.text.x = element_blank(), legend.position = "none")
p3_violin <- ggplot(mtx2@meta.data, aes(x=orig.ident, y=percent.mt))+geom_violin(fill="coral")+ylab("percent.mt")+xlab("")+
  geom_jitter(size=0.5)+theme(axis.text.x = element_blank(), legend.position = "none")
p1<-ggplot(mtx2@meta.data, aes(x=orig.ident, y=log10(nFeature_RNA)))+geom_violin(fill="coral")+ylab("log10(nGene)")+xlab("")+
  geom_jitter(size=0.5)+theme(axis.text.x = element_blank(), legend.position = "none")
p2<-ggplot(mtx2@meta.data, aes(x=orig.ident, y=log10(nCount_RNA)))+geom_violin(fill="coral")+ylab("log10(nUMI)")+xlab("")+
  geom_jitter(size=0.5)+theme(axis.text.x = element_blank(), legend.position = "none")

```

```{r sample_7, message=FALSE,warning = FALSE,include=FALSE}
mtx2<-NormalizeData(mtx2,assay = "RNA");gc()
mtx2 <- SCTransform(mtx2, vars.to.regress = "percent.mt", verbose = FALSE)
mtx2 <- RunPCA(object = mtx2,verbose = FALSE)
mtx2 <- FindNeighbors(mtx2, dims = 1:10, verbose = FALSE)
mtx2 <- FindClusters(mtx2, verbose = FALSE)
mtx2 <- RunUMAP(mtx2, dims = 1:10, verbose = FALSE)
# define the expected number of doublet cells. 
nExp <- round(ncol(mtx2) * 0.04)
mtx2 <- doubletFinder(mtx2, pN = 0.25, pK = 0.09, nExp = nExp, PCs = 1:10,sct = T)
DF.name = colnames(mtx2@meta.data)[grepl("DF.classification", colnames(mtx2@meta.data))]
sample_doublet_f1<-VlnPlot(mtx2, features = "nFeature_RNA", group.by = DF.name, pt.size = 0.1)+
  theme(axis.title.x = element_blank(),legend.position = "none",
        plot.title = element_text(size = 10))+labs(title = "nGene")
sample_doublet_f2<- DimPlot(mtx2, group.by = DF.name, pt.size = 0.1)+
  theme(axis.title.x = element_blank(),axis.title.y = element_blank(),legend.position = c(.7, .8),
        plot.title = element_text(size = 14))+xlab("DoubletFinder")
umi_mt_corr <- FeatureScatter(mtx2, feature1 = "nCount_RNA", feature2 = "percent.mt") + 
  scale_y_continuous(labels = scientific_format()) +
  scale_x_continuous(labels = scientific_format()) +
  theme(legend.position = "none",
        axis.text.x = element_text(size = 9),
        plot.title = element_text(size = 10))+xlab("nUMI")
```

```{r rna_qc_plot,fig.width=13,fig.height=6}
plot_grid(p1_violin,p2_violin,p3_violin,sample_UMI_Gene,p1, p2, umi_mt_corr,sample_doublet_f1,
                    ncol = 4,align = "h",axis = "x")

```

## Column

### False Positive Well Heatmap

```{r false_pos_on_chip_heatmap_rna,results='asis'}
img_file_path <- paste0(params$outdir,"/rawdata/false_pos_on_chip_heatmap.png") 
if (!file.exists(img_file_path)) {
  cat('<p style="color: red; font-weight: bold;">ERROR: Image file not found at ', img_file_path, '</p>')
} else {
  img_data <- readBin(img_file_path, what = "raw", n = file.info(img_file_path)$size)
  img_base64 <- base64enc::base64encode(img_data)
  img_data_uri <- paste0("data:image/png;base64,", img_base64)
  cat('<div style="text-align: center; margin: 5px 0;">')
  cat(sprintf(
    '<img src="%s" alt="False Pos Well Heatmap" id="false_pos_on_chip_heatmap_rna" style="max-width: 100%%; height: auto; cursor: pointer;  object-fit: contain;display: block; margin: 0 auto;" onclick="openBase64ImageInPopup(\'%s\', \'False Pos Well Heatmap\', \'false_pos_on_chip_heatmap_rna.png\')" />',
    img_data_uri, img_data_uri
  ))
  cat('</div>')
}
#knitr::include_graphics("rawdata/rna_process/FILE/false_pos_on_chip_heatmap.png")
```

### Filtered-cells Reads Num Heatmap

```{r real_reads_num_heatmap_rna,results='asis'}
img_file_path <- paste0(params$outdir,"/rawdata/real_reads_num_heatmap.png") 
if (!file.exists(img_file_path)) {
  cat('<p style="color: red; font-weight: bold;">ERROR: Image file not found at ', img_file_path, '</p>')
} else {

  img_data <- readBin(img_file_path, what = "raw", n = file.info(img_file_path)$size)
  img_base64 <- base64enc::base64encode(img_data)
  img_data_uri <- paste0("data:image/png;base64,", img_base64)
  cat('<div style="text-align: center; margin: 5px 0;">')
  cat(sprintf(
    '<img src="%s" alt="Filtered-cells Reads Num Heatmap" id="real_reads_num_heatmap_rna" style="max-width: 100%%; height: auto; cursor: pointer;  object-fit: contain;display: block; margin: 0 auto;" onclick="openBase64ImageInPopup(\'%s\', \'Filtered-cells Reads Num Heatmap\', \'real_reads_num_heatmap_rna.png\')" />',
    img_data_uri, img_data_uri
  ))
  cat('</div>')
}
#knitr::include_graphics("rawdata/rna_process/FILE/real_reads_num_heatmap.png")
```

## Column

### False Negative Well Heatmap

```{r false_neg_on_chip_heatmap_rna,results='asis'}
img_file_path <- paste0(params$outdir,"/rawdata/false_neg_on_chip_heatmap.png") 
if (!file.exists(img_file_path)) {
  cat('<p style="color: red; font-weight: bold;">ERROR: Image file not found at ', img_file_path, '</p>')
} else {

  img_data <- readBin(img_file_path, what = "raw", n = file.info(img_file_path)$size)
  img_base64 <- base64enc::base64encode(img_data)
  img_data_uri <- paste0("data:image/png;base64,", img_base64)
  cat('<div style="text-align: center; margin: 5px 0;">')
  cat(sprintf(
    '<img src="%s" alt="False Negative Well Heatmap" id="false_neg_on_chip_heatmap_rna" style="max-width: 100%%; height: auto; cursor: pointer;  object-fit: contain;display: block; margin: 0 auto;" onclick="openBase64ImageInPopup(\'%s\', \'False Negative Well Heatmap\', \'false_neg_on_chip_heatmap_rna.png\')" />',
    img_data_uri, img_data_uri
  ))
  cat('</div>')
}
#knitr::include_graphics("rawdata/rna_process/FILE/false_neg_on_chip_heatmap.png")
```

### Non-cells Reads Num Heatmap

```{r non_cell_reads_on_chip_heatmap_rna,results='asis'}
img_file_path <- paste0(params$outdir,"/rawdata/non_cell_reads_on_chip_heatmap.png") 
if (!file.exists(img_file_path)) {
  cat('<p style="color: red; font-weight: bold;">ERROR: Image file not found at ', img_file_path, '</p>')
} else {
  img_data <- readBin(img_file_path, what = "raw", n = file.info(img_file_path)$size)
  img_base64 <- base64enc::base64encode(img_data)
  img_data_uri <- paste0("data:image/png;base64,", img_base64)
  cat('<div style="text-align: center; margin: 5px 0;">')
  cat(sprintf(
    '<img src="%s" alt="Non-cells Reads Num Heatmap" id="non_cell_reads_on_chip_heatmap_rna" style="max-width: 100%%; height: auto; cursor: pointer; object-fit: contain; display: block; margin: 0 auto;" onclick="openBase64ImageInPopup(\'%s\', \'Non-cells Reads Num Heatmap\', \'non_cell_reads_on_chip_heatmap_rna.png\')" />',img_data_uri, img_data_uri
  ))
  cat('</div>')
}
#knitr::include_graphics("rawdata/rna_process/FILE/non_cell_reads_on_chip_heatmap.png")
```

# [Typical Markers]{.my-rna}

## Column {.tabset}

### Epithelial

```{r Epithelial, error=TRUE, message=FALSE,warning = FALSE,fig.width=10,fig.height=8}
p1 <-FeaturePlot(object = mtx2,features = Epi_genes, coord.fixed=1,ncol=4,pt.size = 0.1 )&
  theme(
    axis.text.x = element_blank(),  
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),  
    axis.title.y = element_blank()
  )

p2 <-DimPlot(mtx2, reduction = "umap", label = TRUE)+DotPlot(mtx2, features = unique(Epi_genes),assay='RNA',group.by = "seurat_clusters")  + coord_flip()
plot_grid(p1,p2,nrow=2)
```

### Fibroblast

```{r Fibroblast, error=TRUE, message=FALSE,warning = FALSE,fig.width=10,fig.height=8}
p1 <-FeaturePlot(object = mtx2,features = Fibro_genes, coord.fixed=1,ncol=4,pt.size = 0.1)&
  theme(
    axis.text.x = element_blank(),  
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),  
    axis.title.y = element_blank()
  )

p2 <-DimPlot(mtx2, reduction = "umap", label = TRUE)+DotPlot(mtx2, features = unique(Fibro_genes),assay='RNA',group.by = "seurat_clusters")  + coord_flip()
plot_grid(p1,p2,nrow=2)
```

### Endothelial

```{r Endothelial, error=TRUE, message=FALSE,warning = FALSE,fig.width=10,fig.height=8}
p1 <-FeaturePlot(object = mtx2,features = Endo_genes, coord.fixed=1,ncol=4,pt.size = 0.1)&
  theme(
    axis.text.x = element_blank(),  
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),  
    axis.title.y = element_blank()
  )

p2 <- DimPlot(mtx2, reduction = "umap", label = TRUE)+DotPlot(mtx2, features = unique(Endo_genes),assay='RNA',group.by = "seurat_clusters") + coord_flip()
plot_grid(p1,p2,nrow=2)
```

### Adipocytes

```{r Adipocytes, error=TRUE, message=FALSE,warning = FALSE,fig.width=10,fig.height=8}
p1 <-FeaturePlot(object = mtx2,features = Adipo_genes, coord.fixed=1,ncol=4,pt.size = 0.1)&
  theme(
    axis.text.x = element_blank(),  
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),  
    axis.title.y = element_blank()
  )

p2 <-DimPlot(mtx2, reduction = "umap", label = TRUE)+DotPlot(mtx2, features = unique(Adipo_genes),assay='RNA',group.by = "seurat_clusters") + coord_flip()
plot_grid(p1,p2,nrow=2)
```

### Immune Cells

```{r Cells, error=TRUE, message=FALSE,warning = FALSE,fig.width=10,fig.height=8}
p1 <-FeaturePlot(object = mtx2,features = Immun_genes, coord.fixed=1,ncol=4,pt.size = 0.1)&
  theme(
    axis.text.x = element_blank(),  
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),  
    axis.title.y = element_blank()
  )

p2 <-DimPlot(mtx2, reduction = "umap", label = TRUE)+DotPlot(mtx2, features = unique(Immun_genes),assay='RNA',group.by = "seurat_clusters") + coord_flip()
plot_grid(p1,p2,nrow=2)
```

### T-Cells

```{r T_Cells, error=TRUE, message=FALSE,warning = FALSE,fig.width=10,fig.height=8}
p1 <-FeaturePlot(object = mtx2,features = Tcell_genes, coord.fixed=1,ncol=4,pt.size = 0.1)&
  theme(
    axis.text.x = element_blank(),  
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),  
    axis.title.y = element_blank()
  )

p2 <-DimPlot(mtx2, reduction = "umap", label = TRUE)+DotPlot(mtx2, features = unique(Tcell_genes),assay='RNA',group.by = "seurat_clusters") + coord_flip()
plot_grid(p1,p2,nrow=2)
```

### NK-Cells

```{r NK_Cells, error=TRUE, message=FALSE,warning = FALSE,fig.width=10,fig.height=8}
p1 <-FeaturePlot(object = mtx2,features = NKill_genes, coord.fixed=1,ncol=4,pt.size = 0.1)&
  theme(
    axis.text.x = element_blank(),  
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),  
    axis.title.y = element_blank()
  )

p2 <-DimPlot(mtx2, reduction = "umap", label = TRUE)+DotPlot(mtx2, features = unique(NKill_genes),assay='RNA',group.by = "seurat_clusters") + coord_flip()
plot_grid(p1,p2,nrow=2)
```

### B-Cells

```{r B_Cells, error=TRUE, message=FALSE,warning = FALSE,fig.width=10,fig.height=8}
p1 <-FeaturePlot(object = mtx2,features = Bcell_genes, coord.fixed=1,ncol=4,pt.size = 0.1)&
  theme(
    axis.text.x = element_blank(),  
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),  
    axis.title.y = element_blank()
  )

p2 <-DimPlot(mtx2, reduction = "umap", label = TRUE)+DotPlot(mtx2, features = unique(Bcell_genes),assay='RNA',group.by = "seurat_clusters") + coord_flip()
plot_grid(p1,p2,nrow=2)
```

### Macrophages

```{r Macrophages, error=TRUE, message=FALSE,warning = FALSE,fig.width=10,fig.height=8}
p1 <-FeaturePlot(object = mtx2,features = Macro_genes, coord.fixed=1,ncol=4,pt.size = 0.1)&
  theme(
    axis.text.x = element_blank(),  
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),  
    axis.title.y = element_blank()
  )

p2 <-DimPlot(mtx2, reduction = "umap", label = TRUE)+DotPlot(mtx2, features = unique(Macro_genes),assay='RNA',group.by = "seurat_clusters") + coord_flip()
plot_grid(p1,p2,nrow=2)
```

### Monocytes

```{r Monocytes, error=TRUE, message=FALSE,warning = FALSE,fig.width=10,fig.height=8}
p1 <-FeaturePlot(object = mtx2,features = Monoc_genes, coord.fixed=1,ncol=4,pt.size = 0.1)&
  theme(
    axis.text.x = element_blank(),  
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),  
    axis.title.y = element_blank()
  )

p2 <-DimPlot(mtx2, reduction = "umap", label = TRUE)+DotPlot(mtx2, features = unique(Monoc_genes),assay='RNA',group.by = "seurat_clusters")  + coord_flip()
plot_grid(p1,p2,nrow=2)
```

### Dendritic

```{r Dendritic, error=TRUE, message=FALSE,warning = FALSE,fig.width=10,fig.height=8}
p1 <-FeaturePlot(object = mtx2,features = Dendr_genes, coord.fixed=1,ncol=4,pt.size = 0.1)&
  theme(
    axis.text.x = element_blank(),  
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),  
    axis.title.y = element_blank()
  )

p2 <-DimPlot(mtx2, reduction = "umap", label = TRUE)+DotPlot(mtx2, features = unique(Dendr_genes),assay='RNA',group.by = "seurat_clusters") + coord_flip()
plot_grid(p1,p2,nrow=2)
```

### Myeloid

```{r Myeloid, error=TRUE, message=FALSE,warning = FALSE,fig.width=10,fig.height=8}
p1 <-FeaturePlot(object = mtx2,features = myeloid_genes, coord.fixed=1,ncol=4,pt.size = 0.1)&
  theme(
    axis.text.x = element_blank(),  
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),  
    axis.title.y = element_blank()
  )

p2 <-DimPlot(mtx2, reduction = "umap", label = TRUE)+DotPlot(mtx2, features = unique(myeloid_genes),assay='RNA',group.by = "seurat_clusters")  + coord_flip()
plot_grid(p1,p2,nrow=2)
```

### Red Blood Cells

```{r Red_Blood_Cells, error=TRUE, message=FALSE,warning = FALSE,fig.width=10,fig.height=8}
p1 <-FeaturePlot(object = mtx2,features = RedBld_genes, coord.fixed=1,ncol=4,pt.size = 0.1)&
  theme(
    axis.text.x = element_blank(),  
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),  
    axis.title.y = element_blank()
  )

p2 <-DimPlot(mtx2, reduction = "umap", label = TRUE)+DotPlot(mtx2, features = unique(RedBld_genes),assay='RNA',group.by = "seurat_clusters")  + coord_flip()
plot_grid(p1,p2,nrow=2)
```

## Column {data-width="500"}

```{r find_top10_markers, message=FALSE,warning = FALSE}
#library(future)
#plan("multiprocess", workers = 60)
sample_all_cluster.markers <- FindAllMarkers(object = mtx2, only.pos = TRUE)
#write.csv(bcap.markers,paste0(odir,"/selec_markers_17clusters.csv"))
sample_all_cluster.markers %>% group_by(cluster) %>% dplyr::filter(avg_log2FC > 1) %>% slice_head(n = 10) %>% ungroup() -> sample_top10
```

### nUMI UMAP

```{r nUMI_UMAP}
FeaturePlot(mtx2, features ="nCount_RNA")+ 
  theme(axis.title.x = element_blank(),axis.title.y = element_blank(),legend.position = "right",legend.text = element_text(size=12))+labs(title="nUMI")

```

### nGene UMAP

```{r nGene_UMAP}
FeaturePlot(mtx2, features ="nFeature_RNA")+ 
  theme(axis.title.x = element_blank(),axis.title.y = element_blank(),legend.position = "right",legend.text = element_text(size=12))+labs(title="nGene")
```

### DoubletFinder

```{r rna_umap}
sample_doublet_f2
```

## Column {data-width="800"}

### Marker genes for UMAP clusters {data-height="1200"}

```{r top10_heatmap, message=FALSE,warning = FALSE,fig.height=10}
DoHeatmap(object = mtx2,
          features = sample_top10$gene
         )+ theme (axis.text.y = element_text(size=8),legend.text = element_text(size=8))

```

# [CNV Infering]{.my-rna}

## Column

```{r copykat,message=FALSE,warning = FALSE,cache=TRUE,fig.width=3,fig.height=2.5,include=FALSE}
#copykat<-copykat(rawmat=GetAssayData(object = mtx2,slot = "counts"), id.type="S", ngene.chr=5, win.size=25,KS.cut=0.1, sam.name=params$id, distance="euclidean", n.cores=20)
```

### copykat infering CNV

```{r copykat_infering_CNV,results='asis'}
img_file_path <- paste0(params$outdir,"/",params$id,"_copykat_heatmap.jpeg") 
if (!file.exists(img_file_path)) {
  cat('<p style="color: red; font-weight: bold;">ERROR: Image file not found at ', img_file_path, '</p>')
} else {

  img_data <- readBin(img_file_path, what = "raw", n = file.info(img_file_path)$size)
  img_base64 <- base64enc::base64encode(img_data)
  img_data_uri <- paste0("data:image/png;base64,", img_base64)
  cat('<div style="text-align: center; margin: 5px 0;">')
  cat(sprintf(
    '<img src="%s" alt="copykat infering CNV" id="copykat_infering_CNV" style="max-width: 100%%; height: auto; cursor: pointer; display: block; margin: 0 auto;" onclick="openBase64ImageInPopup(\'%s\', \'copykat infering CNV\', \'copykat_infering_CNV.png\')" />',img_data_uri, img_data_uri
  ))
  cat('</div>')

}
#knitr::include_graphics(paste0(params$id,"_copykat_heatmap.jpeg"))
```

```{r copycat_umap,fig.width=6,fig.height=6,eval=TRUE}
copy_mtx2<-read.table(paste0(params$id,"_copykat_prediction.txt"),header=T)
mtx2_meta_data<-tibble::rownames_to_column(mtx2@meta.data,"cell.names")
mtx2_meta_data_2<-merge(mtx2_meta_data,copy_mtx2,by="cell.names",all=TRUE)
mtx2@meta.data <- mtx2_meta_data_2
my_cluster_colors <- c()

if("diploid" %in% mtx2@meta.data$copykat.pred)
{
    my_cluster_colors <- c(my_cluster_colors,"Diploid"="#1B9E77")
}
if("nondiploid" %in% mtx2@meta.data$copykat.pred || "aneuploid" %in% mtx2@meta.data$copykat.pred)
{
      my_cluster_colors <- c(my_cluster_colors,"Aneuploid"="#D95F02")
}
if("c1:diploid:low.conf" %in% mtx2@meta.data$copykat.pred)
{
      my_cluster_colors <- c(my_cluster_colors,"cluster1"="#1B9E77")
}
if("c2:aneuploid:low.conf" %in% mtx2@meta.data$copykat.pred)
{
    my_cluster_colors<-c(my_cluster_colors,"cluster2"="#D95F02")
}
if(sum(is.na(mtx2@meta.data$copykat.pred ))>0)
{
    my_cluster_colors = c(my_cluster_colors,"Undefined"="#66A61E")
}
mtx2@meta.data$copykat.pred <- case_when(
  mtx2@meta.data$copykat.pred == "diploid" ~"Diploid",
  mtx2@meta.data$copykat.pred == "nondiploid" ~"Aneuploid",
  mtx2@meta.data$copykat.pred == "aneuploid" ~"Aneuploid",
   mtx2@meta.data$copykat.pred == "c1:diploid:low.conf"~"cluster1",
   mtx2@meta.data$copykat.pred == "c2:aneuploid:low.conf"~"cluster2",
  TRUE ~"Undefined")
rownames(mtx2@meta.data)<-mtx2@meta.data$cell.names

# old color set up includes even when you don't have multiples
#c("Diploid"="#1B9E77","Aneuploid"="#D95F02","Undefined"="#66A61E","cluster1"="#1B9E77","cluster2"="#D95F02")
```

## Column

### UMI Count Umap and Gene Count Umap

```{r sample_34, message=FALSE,warning = FALSE,fig.width=10,fig.height=4}
p1 <-DimPlot(mtx2,reduction = "umap",group.by = "copykat.pred",cols = my_cluster_colors)+ 
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        legend.position = "bottom",
        legend.text = element_text(size=10))
p2 <- FeaturePlot(mtx2, features ="nCount_RNA")+ labs(title="nUMI")+
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        legend.position = "bottom",
        legend.text = element_text(size=10),
        legend.key.width = unit(1.5, "cm"),    
        legend.key.height = unit(0.5, "cm"))
p3 <- FeaturePlot(mtx2, features ="nFeature_RNA")+ labs(title="nGene")+
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        legend.position = "bottom",
        legend.text = element_text(size=10),
        legend.key.width = unit(1.5, "cm"),    
        legend.key.height = unit(0.5, "cm"))
plot_grid(p1,p2,p3,nrow = 1)
```

# [Coassay]{.my-coassay}

## Column {data-width="500"}

```{r venn,include=FALSE,message=FALSE,warning = FALSE}
new_pal = c("#CC0C00B2", "#5C88DAB2", "#84BD00B2", "#FFCD00B2", "#7C878EB2", "#00B5E2B2", "#00AF66B2", "#D2AF81B2",
            "#FD7446B2", "#46732EB2", "#C1395E", "#E07B42","#D4A2D9", "#8E72D5","#C0EDB9", "#364E4F", "#8EE5EE",
            "#FFA500", "#458B00", "#CD6090", "#FFAEB9", "#90EE90", "#5f9EA0", "#E6E6FA", "#8B7E66", "#a6cee3","#1f78b4",
            "#b2df8a","#33a02c","#fb9a99","#e31a1c","#fdbf6f","#ff7f00","#cab2d6","#6a3d9a","#ffff99","#b15928",
            "#3d348b","#f72585","#7678ed","#9b5de5","#2ec4b6", "#fff1d0")
new_pal2 <- new_pal
pro_name <- params$id
pro_name_r <- paste0(pro_name, "_rna")
pro_name_d <- paste0(pro_name, "_dna")
###---DNA--------#########
source("/sibcb3/wangkailelab3/jinyu/dr-analysis/dr-analysis/0_wdr_functions.R")

varbin_mtx <- readRDS(paste0(params$outdir,"/final_result/",params$id,"_copykit_obj.rds"))

varbin_mtx_filter <- filterCells(varbin_mtx,resolution = 0.8)
filt_cells_names <- as.data.frame(varbin_mtx_filter@colData) %>% dplyr::filter(filtered=="kept") %>% select(sample)
varbin_mtx_filter <- varbin_mtx[,filt_cells_names$sample]
#----------UMAP-----
set.seed(31)
varbin_mtx_filter_log <- varbin_mtx_filter
umap_data <-  data.frame(uwot::umap(log(t(varbin_mtx_filter_log@assays@data$segment_ratios+1e-3)),
                                    metric = "manhattan", min_dist = 0.1, spread = 3, n_neighbors = 20))
umap_data2 <- umap_data %>% dplyr::rename("UMAP_1" = "X1", "UMAP_2" = "X2")
varbin_mtx_filter_log@int_colData$reducedDims <- umap_data2
varbin_mtx_filter_log@colData <- cbind(varbin_mtx_filter_log@colData, umap_data2)
#---subclone---
set.seed(17)
hdb_data <- dbscan::hdbscan(umap_data2, minPts = nrow(umap_data2)*0.015)
subclones <- paste0("c", as.character(hdb_data$cluster))
varbin_mtx_filter_log@colData <- cbind(varbin_mtx_filter_log@colData, subclones)
num_clst <- length(unique(varbin_mtx_filter_log@colData$subclones))
if(length(new_pal) >= num_clst) {
  new_pal_tmp = new_pal
} else {
  new_pal_tmp = hue_pal()(num_clst)
}
subclones_name <- table(subclones) %>% as.data.frame()
varbin_mtx_filter_log@colData$subclones <- factor(varbin_mtx_filter_log@colData$subclones,
                                                  levels = subclones_name$subclones)
#singler
load("/sibcb1/wangkailelab1/pengjinyu/tools/bin/refdata.RData")
norm_count_aligned <- GetAssayData(mtx2, slot = "data")
clusters_aligned <- mtx2@meta.data[colnames(norm_count_aligned), "seurat_clusters", drop=TRUE] 

if (ncol(norm_count_aligned) != length(clusters_aligned)) {
    stop("Length mismatch after alignment!")
}

pred.results <- SingleR(
    test = norm_count_aligned,
    ref = refdata,
    labels = refdata$label.main,
    clusters = clusters_aligned, 
    assay.type.test = "logcounts",
    assay.type.ref = "logcounts"
)

pred.results <- pred.results %>% as.data.frame()%>% rownames_to_column("seurat_clusters")
meta_data <- mtx2@meta.data %>% as.data.frame() %>% left_join(.,pred.results,by="seurat_clusters")
mtx2@meta.data$singleR_label <- meta_data$labels
new.clusters <- pred.results$labels
names(new.clusters) <- levels(mtx2)
mtx2 <- RenameIdents(mtx2, new.clusters)


wafergen <- read.csv("/sibcb3/wangkailelab3/jinyu/demultiplex/index/wafergen_list_DNA_RNA.txt",sep = " ")
wafergen$rna_name <- paste0(wafergen$rna_n7_cr,wafergen$rna_s5)
name_dna <- varbin_mtx_filter_log@colData %>%
  as.data.frame() %>%
  select(sample,subclones) %>%
  mutate(
    split_sample = stringr::str_split(sample, "_"),
    Row = sapply(split_sample, `[`, 3),
    Col = sapply(split_sample, `[`, 4)
  )%>%mutate(Row=as.numeric(gsub("r","",Row)) ,Col=as.numeric(gsub("c","",Col)))%>% left_join(wafergen, by = c("Row","Col")) %>% 
  select(sample,rna_name,subclones)


DNA_rna_meta <- mtx2@meta.data %>% rownames_to_column()  %>% 
  dplyr::left_join(name_dna, ., by = c("rna_name" = "rowname")) #%>% 
#dplyr::mutate(cell_type = plyr::mapvalues(seurat_clusters, from = levels(obj_rna@meta.data$seurat_clusters), to = rna_celltype))

varbin_mtx_filter_log@colData <- cbind(varbin_mtx_filter_log@colData, DNA_rna_meta)
# VennDiagram 
dna <- varbin_mtx_filter_log@colData$sample
rna <- mtx2@meta.data$orig.ident
dna_rna <- as.data.frame(varbin_mtx_filter_log@colData) %>% filter(!is.na(orig.ident))
area1 <- length(dna) 
area2 <- length(rna) 
cross_area <- length(dna_rna$sample) 
venn.plot <- draw.pairwise.venn(area1 = area1, area2 = area2, cross.area = cross_area,
                                category = c("DNA", "RNA"),
                                fill = c("skyblue", "pink"), 
                                alpha = 0.5, 
                                cex = 1, 
                                cat.cex = 1, 
                                cat.col = c("black", "black")) 

```

```{r heatmap_rna_dna,include=FALSE,message=FALSE}
mtx_srt <- as.data.frame(varbin_mtx_filter_log@colData) %>%
  arrange(factor(subclones, levels = levels(varbin_mtx_filter_log@colData$subclones)))
ht_mtx <- log2(t(varbin_mtx_filter_log@assays@data$segment_ratios))[mtx_srt$sample,]

#----annotation bar----
anno_mtx <- mtx_srt %>% mutate(rna_clst0 = as.character(singleR_label)) %>% mutate(rna_clst = ifelse(is.na(rna_clst0), "RNA_miss",rna_clst0)) %>%
  dplyr::select(c("subclones", "rna_clst"))
rownames(anno_mtx) <- NULL
cluster_num <- length(unique(anno_mtx$subclones))
clst_col <- new_pal[1:length(unique(anno_mtx$subclones))]
names(clst_col) <- levels(varbin_mtx_filter_log@colData$subclones)
rna.cluster.col <- c("#D89000","#da88dc","#00BD5F","#9590FF","#F8766D","#54278f","#BF4000","#3969AC","grey90","grey60","grey70","grey80","#CAB2D1","red")

rna_col <- c(rna.cluster.col[1:length(unique(meta_data$labels))],"grey90")
names(rna_col)<-c(unique(meta_data$labels),"RNA_miss")

breaks = c(-1,0,1)
col_vec = circlize::colorRamp2(breaks =breaks, c("dodgerblue3", "white", "firebrick3"))

ha_row2_new=rowAnnotation(df = anno_mtx, col = list(subclones=clst_col,rna_clst=rna_col),
                      show_annotation_name = F, simple_anno_size = unit(0.7, "cm"))

ht_combine <- Heatmap(as.matrix(ht_mtx), cluster_columns = FALSE, border = TRUE, cluster_rows = FALSE, show_row_dend = FALSE,
        row_split = anno_mtx$subclones ,
        name = "scheatmap", show_row_names = F, show_column_names = F, row_title = paste0(nrow(ht_mtx), " single cells"),
        use_raster = T, raster_quality = 5, col = col_vec,
        heatmap_legend_param = list(title = "Log2 (Ratio)",
                                    title_gp = gpar(fontsize = 12, fontface = "bold"), labels_gp = gpar(fontsize = 12)),
        top_annotation = chr_bar, left_annotation = ha_row2_new)

# RNA data UMAP----
RNA_dna_subclone <- mtx2@meta.data %>% rownames_to_column()  %>% 
  dplyr::left_join( .,name_dna, by = c("rowname" = "rna_name")) %>%
  dplyr::select(subclones) %>%
  mutate_if(is.factor, as.character) %>%
  mutate_all(~ replace(., is.na(.), "unsigned"))

mtx2@meta.data$subclone <- RNA_dna_subclone$subclones

#---map DNA clusters to RNA UMAP----#####
plist <- list()
j <- 1
mtx2@meta.data$subclone <- factor(mtx2@meta.data$subclone,levels = c(levels(name_dna$subclones),"unsigned"))
my_clones <- levels(mtx2@meta.data$subclone)
for (i in my_clones[1:length(my_clones)]) {
  #i <- "c0" 
  obj_rna_newid_tmp <- mtx2
  obj_rna_newid_tmp@meta.data <- obj_rna_newid_tmp@meta.data %>% mutate(subclone4 = ifelse(subclone == i, as.character(subclone), "unsel"))
  
  mut_cell_df <- obj_rna_newid_tmp@meta.data[,c("seurat_clusters","subclone")] %>% table() %>% as.data.frame() %>% dplyr::filter(Freq >0 & Freq <3)
  mut_cell_chr <-  mut_cell_df %>% dplyr::filter(subclone == i) %>% pull(seurat_clusters) %>% as.character()
  obj_rna_newid_tmp@meta.data <- obj_rna_newid_tmp@meta.data %>% mutate(subclone4 = ifelse((subclone4 == i & seurat_clusters %in% mut_cell_chr), "unsel", subclone4))
  
  p1 <- DimPlot(obj_rna_newid_tmp, reduction = "umap", group.by = "subclone4", pt.size = 1, order = i) + 
    scale_color_manual(values = c("grey90",new_pal[j])) + ggtitle("")+ theme_bw() +
    theme(panel.grid.major = element_blank(),panel.grid.minor = element_blank(),panel.background = element_rect(colour = "black", linewidth=0.3), 
          axis.ticks = element_blank(),axis.text = element_blank(), axis.title = element_text(size = 8),legend.position = "bottom",aspect.ratio=1)
  plist[[j]] <- p1
  j <- j +1
}

if(cluster_num >6){
  p_list1 <- plot_grid(plotlist = plist, ncol = 6)
  if((length(my_clones) %%6)>0){
    fig_height <- (length(my_clones) %/%6)+1
    plot_width <- 6
  }else{
    fig_height <- (length(my_clones) %/%6)
    plot_width <- 6
  }
  
}else{
  p_list1 <- plot_grid(plotlist = plist, ncol = cluster_num)
  fig_height <- 1
  plot_width <- cluster_num
}

ggsave(filename =paste0(params$outdir, "/final_result/figures/DNA_subclone_on_RNA_UMAP.png"), plot = p_list1, width = plot_width*3, height = fig_height*3)
#ggsave(filename =paste0(params$outdir, "/final_result/figures/DNA_subclone_on_RNA_UMAP.png"), plot = p_list1, width = 18, height = 8)
# bar plot

celltype <- unique(pred.results$labels)
if(length(celltype) >1){
  meta_con <- varbin_mtx_filter_log@colData %>% as.data.frame() %>% dplyr::select(subclones, singleR_label) %>% 
  dplyr::group_by(subclones, singleR_label) %>% dplyr::summarise(n=n()) %>% spread(singleR_label, n) %>% dplyr::select(-`<NA>`)
  meta_con[is.na(meta_con)] <- 0
  meta_con_sum <- meta_con %>% column_to_rownames(var = "subclones") %>% mutate(mysum=apply(., 1, sum))
  meta_con_p <- meta_con_sum/meta_con_sum$mysum 
  meta_con_p2 <- meta_con_p %>% dplyr::select(-mysum)
  meta_con_p2_prop <- as.data.frame(t(apply(meta_con_p2, 1, function(x) x/sum(x))))
  meta_con_p2_prop$rowname <- rownames(meta_con_p2_prop)
  meta_con_p2_long <- tidyr::pivot_longer(
    meta_con_p2_prop,
    cols = -rowname,
    names_to = "variable",
    values_to = "proportion")
  bar_plot <- ggplot(meta_con_p2_long, aes(x = rowname, y = proportion, fill = variable)) +
    geom_bar(stat = "identity") +scale_fill_manual(values = rna_col)+
    labs(y = "Proportion",
         x = "Subclones",
       fill = "Cell Type") +
    theme(legend.text = element_text(size = 5),
          legend.title =element_text(size = 6),
          axis.title.x = element_text(size = 6),
          axis.title.y = element_blank(),
          axis.text.x = element_text(angle = 90,size = 6),
          axis.text.y = element_text(size = 6))
} else{
  meta_con <- varbin_mtx_filter_log@colData %>% as.data.frame() %>% dplyr::select(subclones, singleR_label) %>% 
  dplyr::group_by(subclones, singleR_label) %>% dplyr::summarise(n=n()) %>% spread(singleR_label, n) %>% dplyr::select(-`<NA>`)
  meta_con[is.na(meta_con)] <- 0
  meta_con_sum <- meta_con %>% column_to_rownames(var = "subclones") %>% mutate(mysum=apply(., 1, sum))
  meta_con_p <- meta_con_sum/meta_con_sum$mysum 
  meta_con_p2 <- meta_con_p %>% dplyr::select(-mysum)  
  meta_con_p2$rowname <- rownames(meta_con_p2)
  meta_con_p2_long <- tidyr::pivot_longer(
    meta_con_p2,
    cols = -rowname,
    names_to = "variable",
    values_to = "proportion")
  bar_plot <- ggplot(meta_con_p2_long, aes(x = rowname, y = proportion, fill = variable)) +
    geom_bar(stat = "identity") +scale_fill_manual(values = rna_col)+
    labs(y = "Proportion",
         x = "Subclones",
       fill = "Cell Type") +
    theme(legend.text = element_text(size = 5),
          legend.title =element_text(size = 6),
          axis.title.x = element_text(size = 6),
          axis.title.y = element_blank(),
          axis.text.x = element_text(angle = 90,size = 6),
          axis.text.y = element_text(size = 6))
}



ggsave(filename =paste0(params$outdir, "/final_result/figures/celltype_propotion_barplot.png"), plot = bar_plot, width = 5, height = 3)
```

### DNA-RNA Overlap

```{r venn_plot,fig.width=5,fig.height=3}
grid.draw(venn.plot)
#plot_grid(venn.plot,bar_plot,ncol = 2)
```

### Annotation by SingleR

```{r umap_new_cluster,fig.width=6}
DimPlot(mtx2, reduction = "umap", label = TRUE)+scale_color_manual(values = rna_col)
```

## Column {.tabset}

### DNA-RNA Combined Heatmap {data-width="1500"}

```{r heatmap,fig.width=10,fig.height=12}
draw(ht_combine,heatmap_legend_side = "right")

```

### DNA clusters on RNA UMAP {data-width="1500"}

```{r DNA_clusters_to_RNA_UMAP,results='asis'}
img_file_path <- paste0(params$outdir,"/final_result/figures/DNA_subclone_on_RNA_UMAP.png") 
if (!file.exists(img_file_path)) {
  cat('<p style="color: red; font-weight: bold;">ERROR: Image file not found at ', img_file_path, '</p>')
} else {
  img_data <- readBin(img_file_path, what = "raw", n = file.info(img_file_path)$size)
  img_base64 <- base64enc::base64encode(img_data)
  img_data_uri <- paste0("data:image/png;base64,", img_base64)
  cat('<div style="text-align: center; margin: 5px 0;">') 
  cat(sprintf(
    '<img src="%s" alt="DNA clusters on RNA UMAP" id="DNA_subclone_on_RNA_UMAP" style="max-width: 100%%; height: auto; cursor: pointer; display: block; margin: 0 auto;" onclick="openBase64ImageInPopup(\'%s\', \'DNA clusters on RNA UMAP\', \'DNA_subclone_on_RNA_UMAP.png\')" />',img_data_uri, img_data_uri
  ))
  cat('</div>')
}
#knitr::include_graphics("final_result/figures/DNA_subclone_on_RNA_UMAP.png")

```

### Celltype Proportion {data-width="1500"}

```{r celltype barplot,results='asis'}
img_file_path <- paste0(params$outdir,"/final_result/figures/celltype_propotion_barplot.png") 
if (!file.exists(img_file_path)) {
  cat('<p style="color: red; font-weight: bold;">ERROR: Image file not found at ', img_file_path, '</p>')
} else {
  img_data <- readBin(img_file_path, what = "raw", n = file.info(img_file_path)$size)
  img_base64 <- base64enc::base64encode(img_data)
  img_data_uri <- paste0("data:image/png;base64,", img_base64)
  cat('<div style="text-align: center; margin: 5px 0;">')
  cat(sprintf(
    '<img src="%s" alt="Celltype Proportion" id="celltype_propotion_barplot" style="max-width: 100%%; height: auto; cursor: pointer; display: block; margin: 0 auto;" onclick="openBase64ImageInPopup(\'%s\', \'Celltype Proportion\', \'celltype_propotion_barplot.png\')" />',img_data_uri, img_data_uri
  ))
  cat('</div>')
}
#knitr::include_graphics("final_result/figures/celltype_propotion_barplot.png")
```

# [About]{.my-about}

## Column

### Dashboard

This dashboard was developed by Jinyu PENG.

### DNA filter

DNA filter is using the following parameters：

-   resolution: 0.8

### Clustering

Clustering of the data is being performed with UMAP dimension reduction followed by R phenograph.

UMAP is using the following parameters:

-   Seed: 31
-   Distance: Manhattan
-   min_dist: 0.1
-   n_neighbors: 20 (if n_cells \< 20 then k = n_cells - 1)

RPhenograph is using the following parameters:

-   k: 20 (if n_cells \< 20 then k = n_cells - 1)

### Ploidy Inference

Ploidy inference is using the fourier transform method developed by Alexander Davis. For more information please refer to the following publication. Ref: Soon

### Phylogenetics

Phylogenetics is using the Neighbor-Joining method and the distance matrix is calculated using manhattan distance.

### Disclaimer

copykat is the Navin lab tool to inferCNV. please refer to our published paper "Delineating copy number and clonal substructure in human tumors from single cell transcriptomes" (<https://www.nature.com/articles/s41587-020-00795-2>) for more detail.

DoubletFinder uses gene expression features to predict doublets in scRNA-seq ([https://www.cell.com/cell-systems/fulltext/S2405-4712(19)30073-0](https://www.cell.com/cell-systems/fulltext/S2405-4712(19)30073-0){.uri})

## Column

### Packages

The following packages were used in this dashboard:

-   flexdashboard
-   tidyverse
-   here
-   fs
-   cowplot
-   DT
-   shiny
-   janitor
-   ape
-   ggtree
-   ggsci
-   plotly
-   amap
-   paletteer
-   scales
-   uwot
-   dbscan
-   ComplexHeatmap
-   flsa
-   Rphenograph
-   Seurat
-   copykat
-   scquantum
